### PR TITLE
[SYCL] Support or diagnose use of namespace std types as kernel type names

### PR DESCRIPTION
**Pull request by jtmott-intel**
This is a continuation of #1579. Sorry I couldn't commit to the branch named in that PR.

*This* PR has premanandrao's original commit plus one of mine to address the comments in 1579 -- [specifically this conversation](https://github.com/intel/llvm/pull/1579#pullrequestreview-406823625).

I removed the changes to the generated header and added a diagnostic to make nullptr_t in kernel names an error. It's possible to make nullptr_t work like int works, but sounds like we don't want to encourage using builtin types as kernel names. The diagnostic message is open to wordsmithing.

### CODE REVIEW

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (Fznamznon) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10838]: Just wondering, why we disallow only nullptr_t?
Your PR message says " sounds like we don't want to encourage using builtin types as kernel names. "

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:1901]: I spend some time looking into these changes, and I don't get why we didn't start emitting "kernel name is missing" on each unnamed lambda, because before your changes this diagnostic was under `if (!UnnamedLambdaSupport)` and it looked correct.

**Comment by erichkeane** [comment]
I disagree with this approach. While Roland's comment is 'Because there is no promise that std types work as lambda name', that only applies as the top level item.  a kernel named 

    <std::string>

SHOULD be illegal, if we're going to implement Roland's comment, we should cover EVERY type in std.

However, 

    <SomeTemplate<std::string>>
    or
    <SomeTemplate<std::nullptr_t>>

the std type is NOT the kernel name, an instance of SomeTemplate is.  We should have to support this case.

I believe we'd discussed before that we should just serialize std::nullptr_t as decltype(nullptr) to make this work.

**Comment by premanandrao** [comment]
> I disagree with this approach. While Roland's comment is 'Because there is no promise that std types work as lambda name', that only applies as the top level item. a kernel named
> 
> ```
> <std::string>
> ```
> 
> SHOULD be illegal, if we're going to implement Roland's comment, we should cover EVERY type in std.
> 
> However,
> 
> ```
> <SomeTemplate<std::string>>
> or
> <SomeTemplate<std::nullptr_t>>
> ```
> 
> the std type is NOT the kernel name, an instance of SomeTemplate is. We should have to support this case.
> 
> I believe we'd discussed before that we should just serialize std::nullptr_t as decltype(nullptr) to make this work.

Agreed with the second part - only top-level should be disallowed.
About the first part - did Roland mean 'all' std types or just the fundamental ones like size_t, ptrdiff_t etc.?

**Comment by erichkeane** [comment]
> 
> 
> > I disagree with this approach. While Roland's comment is 'Because there is no promise that std types work as lambda name', that only applies as the top level item. a kernel named
> > ```
> > <std::string>
> > ```
> > 
> > 
> > SHOULD be illegal, if we're going to implement Roland's comment, we should cover EVERY type in std.
> > However,
> > ```
> > <SomeTemplate<std::string>>
> > or
> > <SomeTemplate<std::nullptr_t>>
> > ```
> > 
> > 
> > the std type is NOT the kernel name, an instance of SomeTemplate is. We should have to support this case.
> > I believe we'd discussed before that we should just serialize std::nullptr_t as decltype(nullptr) to make this work.
> 
> Agreed with the second part - only top-level should be disallowed.
> About the first part - did Roland mean 'all' std types or just the fundamental ones like size_t, ptrdiff_t etc.?

I read it as 'all'... @rolandschulz can you clarify?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (jtmott-intel) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10838]: nullptr_t is special because builtin type names are (typically) accessible without a header, but nullptr_t is a builtin type that isn't known by any name until we include a header (or forward declare). The "don't want to encourage using builtins" comes from [rolandschulz's comment](https://github.com/intel/llvm/pull/1579#pullrequestreview-406823625) -- "Adding extra support for std types is not needed," and, "there is no promise that std types work as lambda name." But I may have misinterpreted his comments, and others asking him for clarification.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:1901]: The original "if" was `if (TD && !UnnamedLambdaSupport)`, and when this evaluates false, then the else block would still use `TD` anyway -- which seems weird. premanandrao's commit refactored the TD check out to `if (!TD) break;`

**Comment by jtmott-intel** [comment]
While we wait for clarification, just adding a couple notes:

- If we use the `decltype(nullptr)` trick, then `std::nullptr_t` would work for \*both\* the top-level kernel name and as a template parameter.
- Some types may never work, not even as template parameters, such as `<SomeTemplate<std::max_align_t>>`, because max_align_t is an anonymous struct. The issue isn't whether it's the top-level kernel name or template parameter. The issue is can we write it out as textual source to the integration header, and that complication isn't limited to std types.

**Comment by rolandschulz** [comment]
Yes I meant all std types. Only types which can be forward declared have to be supported and it isn't valid to forward declare std types.

**Comment by erichkeane** [comment]
>     * Some types may never work, not even as template parameters, such as `<SomeTemplate<std::max_align_t>>`, because max_align_t is an anonymous struct. The issue isn't whether it's the top-level kernel name or template parameter. The issue is can we write it out as textual source to the integration header, and that complication isn't limited to std types.

This point here is particularly annoying... @rolandschulz : what clarification to the SYCL spec needs to me made to allow us to prevent ANY use of namespace-std in kernel names?  Or do you and @mkinsner / @jbrodman /etc already feel that is the case?

**Comment by rolandschulz** [comment]
For the implementation we should just assume that it's OK that nothing in std is allowed. This is clearly the intention. If someone thinks this needs to be clarified in the spec even for 1.2.1 not just upcoming SYCL we can do that.

**Comment by jtmott-intel** [comment]
One of the relevant use cases seems to be a kernel inside a function template, because the function template parameters get reused in the kernel name to keep each name type unique. We do this ourselves in the SYCL handler "fill" function template. The value type to fill is a parameter that gets reused in the kernel name. If we call fill with a `std::complex` number, then if the kernel rejects it, fill rejects it.

**Comment by jtmott-intel** [comment]
Latest commit disallows names from std namespace. `std::nullptr_t` is still special cased since it resolves to a builtin instead of a decl.

**Comment by bader** [comment]
@jtmott-intel, please, resolve CI failures.

**Comment by jtmott-intel** [comment]
All checks have passed.

Here's a quick re-summary.

Originally we were going to emit an error diagnostic for kernel names with `std::nullptr_t`. That type was never guaranteed to work as a kernel name, and indeed the integration header wasn't generating correctly for that type.

Later we decided to disallow all `std` types, since the standard doesn't allow us to forward declare anything in `std`. If ever a `std` type appeared to work in a kernel name, it's because the dragons were merciful.

After implementing this change, several tests broke because they were (indirectly) using `std` types in their kernel names. Most of those occurrences were a kernel inside a template, because the template needs different kernel names for each generated specialization, and when we re-purposed the template's parameters for the kernel name, then any `std` type passed to the template also ended up in the name.

    // example, before this PR
    template<typename T> void test() {
        ...
        cgh.parallel_for<KernelName<T>>( ...
        ...
    }
    test<std::nullptr_t>();       // bad, std::nullptr_t indirectly used in kernel name
    test<std::complex<double>>(); // bad, std::complex indirectly used in kernel name

I updated the impacted tests and made the kernel name its own distinct template parameter so we wouldn't need to re-purpose the other parameters anymore.

    // example, after this PR
    template<typename SpecializationKernelName, typename T> void test() {
        ...
        cgh.parallel_for<SpecializationKernelName>( ...
        ...
    }
    test<class SpecializationKernelName_ABC, std::nullptr_t>();       // ok, kernel name and std::nullptr_t are separate
    test<class SpecializationKernelName_XYZ, std::complex<double>>(); // ok, kernel name and std::complex are separate

This does, however, put more responsibility on the user to pass unique kernel names. If they don't, then existing diagnostics will already tell them.

**Comment by bader** [comment]
> This is a continuation of #1579. Sorry I couldn't commit to the branch named in that PR.

Please, close #1579 if you are not going to merge it.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: I think this should be allowed. The KernelName here is `Templated_kernel_name<std::nullptr_t>` This is not in the 'std' namespace. I do not think there is any spec restriction which says we cannot template a kernel with types from std namespace. 

I think this is why the tests are failing as well. Changing this will probably fix the tests without the need for 'SpecializationKernelName'. I think this change might be breaking valid code

Comment (erichkeane) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: The problem here is that the justification here is the same.  Whether you do kernel_single_task<std::string> or kernel_single_task<some_template<std::string>>, we cannot actually put it in the int-header, because we cannot forward declare std::string.

nullptr_t and std::size_t/ptrdiff_t are less troublesome, since we know they are well defined typedefs (and can just spell them that way), but other types we cannot.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (erichkeane) [inline clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:59]: I think that any exceptions we have to allow std types like size_t and ptrdiff_t, we need to just canonicalize them.  So this should be kernel_single_task<unsigned long long>.  Otherwise we're still violating the same rule as with std::string.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (premanandrao) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: > the standard doesn't allow us to forward declare anything in `std`.

@jtmott-intel, if a Kernel Name is not a std-type itself, but is templated on a std-type instead, like: foo<std::some_type>, what is the case where we need to forward-declare the std-type?   We only need to forward-declare foo, right?

Comment (erichkeane) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: @premanadrao: How would we get away with that?  We cannot spell foo<std::string> without spelling std::string.

Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: In current code, I think std::some_type is also forward declared. There is no check on template parameters IIRC. However, if std header is included prior to integration header, I see no reason why we need to forward declare std::sometype

Comment (premanandrao) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: > @premanadrao: How would we get away with that? We cannot spell foo<std::string> without spelling std::string.

Ah, because the integration header does not necessarily follow the included std header.   Got it.

Comment (erichkeane) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: > In current code, I think std::some_type is also forward declared. 
The C++ standard prohibits doing this.

As I think about it, we could potentially FORCE the int-header to include some of the std:: headers, but that would make compile times even more interesting... Particularly because we don't seem to have a good way to inject the includes into the int-headers after the fact.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (jtmott-intel) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:59]: The complication with `std::size_t` and a couple others like it is by the time we get to `Sema::ConstructOpenCLKernel`, the template argument type already evaluates to the builtin type `unsigned long`, and not convertible to a Decl. At this stage of the compilation, it doesn't seem like I can distinguish between the user spelling "std::size_t" or "unsigned long". Maybe I need an AST visitor to check how the user spelled the kernel name type in source?

Comment (erichkeane) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:59]: I think you misunderstood my comment (and apparently github ate the 'so this should be kernel_single_task&lt;unsigned long&gt;> line) ... I think we need to prohibit any of the classes in the std namespace, but size_t/ptrdiff_t/etc are both typedefs, so just canonicalizing them (aka, desugaring the typedef) should be sufficient.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (rolandschulz) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: We could include exactly those std:: headers which the original source file includes. That shouldn't add anything to the compile time because they should all have proper include guards.

Off topic for this PR (which is focused on std types): This doesn't solve the wider problem. A user can have other library types which aren't fwd-declarable. But for any arbitrary library headers we can't include them from the int-header because we don't know whether they have proper include guards.

**Comment by rbegam** [comment]
please resolve the conflicts as well.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (jtmott-intel) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: There's a lot of interesting ideas here. Including the headers that the original source includes could remove any need for forward declarations and all the problems that go with it. So my next question is, should we abandon this PR? This PR tweaks the existing forward declaration approach.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (jtmott-intel) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:59]: OK. If I understand what you mean, then size_t/ptrdiff_t are already desurgared. A `kernel_single_task<std::size_t>` is written to the integration header as `kernel_single_task<unsigned long>`. That's why those names still work as kernel names.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (rolandschulz) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: Including headers from the int-header will likely cause all kinds of problems. We would effectively change the include order the user wrote. This will mess with e.g. precompiled headers, system headers which use preprocessor switches (e.g. `__USE_FILE_OFFSET64` or `WIN32_LEAN_AND_MEAN`) and user/library headers which redefines macros.

Comment (jtmott-intel) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: OK. If after this discussion, we're deciding to stay with the existing forward declaration approach, then is this PR okay to merge? I'll need fresh approvals from code owners. @elizabethandrews, @Fznamznon, @premanandrao

Comment (erichkeane) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: I still think we have to prohibit everything in namespace 'std', with some allowance for the types where desugaring/canonicalizing results in a non-std type (basically, anything that was typedef'ed to a builtin type).

Comment (jtmott-intel) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: @erichkeane I think that's what we have now in this PR. The [`if (NS->isStdNamespace()) {`](https://github.com/intel/llvm/pull/1963/files#diff-c88a590aca6d5f8e92dd1f8b6ec05a49R2273) prohibits everything in namespace std. Some types such as `std::size_t` desugar to a builtin type such as `unsigned long` and are still allowed to work. Only `nullptr_t` is singled out, since `std::nullptr_t` also desugars to a builtin type but one that we explicitly check for and disallow. Is there a specific case you had in mind that should be changed?

Comment (erichkeane) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: No just being clear.  I wouldn't mind allowing std::nullptr_t as well, since we can just canonicalize it as decltype(nullptr_t) though.

Comment (rolandschulz) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: I don't know what to do here. Things need to be forward declarable for us to do the right thing in the int-header. std types aren't. So we don't really have a choice. But we have codes using std-types in the name. Therefore if we go ahead with this solution we'll break codes. And the workaround shown above (with `SpecializationKernelName_XYZ`) doesn't work for someone who needs to write a kernel which works with any type. They can't provide a custom name for each without asking their users to provide a name. I hope @mkinsner has an idea.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (rolandschulz) [inline clang/lib/Sema/SemaSYCL.cpp:2273]: This whole thing is a non-issue if unnamed lambda compiler option is enabled even if the user provides a kernel name. Because then we don't template on the lambda name type but on the string representation of it. Therefore we don't need to emit an error in that case.

Comment (rolandschulz) [reply clang/lib/Sema/SemaSYCL.cpp:2273]: It seems like we could always do that: Template `KernelInfo` on the string representation of the name rather than the name type. Then we would never need to forward declare the name type.

File: clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp
Comment (rolandschulz) [reply clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp:54]: Just talked with @mkinsner . The intention is that that names and their template arguments are forward declarable. This excludes std types. We think it is OK for us to not support names with std names as template arguments. For those use cases users should use unnamed lambdas.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:2411]: I started experimenting with templating KernelInfo on the string representation of the name, and probably went further down that rabbit hole than I should have before I realized I was reimplementing what the existing unnamed lambda logic does already. And so in this latest commit -- meant to prompt discussion -- I flipped a 0 to a 1 (don't laugh :-P) and turned on the unnamed lambda CLI option by default. This seems to deliver exactly what we've been wanting. KernelInfo is templated on the string representation of the name, nullptr_t and other std names work, and the existing code will already skip forward decls when unnamed lambda is on. I assume, though, that the idea of turning on unnamed lambdas by default was already discussed in some meeting somewhere. Is there some reason it can't be this simple?

@rolandschulz @erichkeane @premanandrao @elizabethandrews

**Comment by erichkeane** [comment]
We shouldn't turn the unnamed lambdas option on by default, simply because it isn't part of the SYCL standard.

THAT SAID, even if we DO, we HAVE to prohibit these types in the non-unnamed lambdas mode, so the original idea of the patch still needs to be implemented. Setting the unnamed lambdas to default doesn't change this.

**Comment by Jeff-Mott-OR** [comment]
Fixing the implementation so std names work seems a pointless exercise if we're going to block std names anyway with diagnostics. I'm trying to find the path that makes everyone happy. :-)

Several folks here expressed a strong desire to have std names work, and the good news is we *can* make them work, and the implementation to do so -- templating KernelInfo on the string representation of the name -- is already in place; we just need to switch it on.

Several folks here also expressed that the intention of the spec is that names and their template arguments be forward declarable, which excludes std types. That implementation is also ready and in this PR.

So now it seems we just need to pick A or B.

Personally I think making std types work is what's ultimately better for *the users*. It's also needed for our own `cl::sycl::handler::fill` to work, since the "fill" function template uses the value type in its kernel name.

But either way, I wouldn't want this PR to keep going in circles, so... do we want A or B?

@rolandschulz @erichkeane @premanandrao @elizabethandrews

**Comment by jtmott-intel** [comment]
Fixing the implementation so std names work seems a pointless exercise if we're going to block std names anyway with diagnostics. I'm trying to find the path that makes everyone happy. :-)

Several folks here expressed a strong desire to have std names work, and the good news is we *can* make them work, and the implementation to do so -- templating KernelInfo on the string representation of the name -- is already in place; we just need to switch it on.

Several folks here also expressed that the intention of the spec is that names and their template arguments be forward declarable, which excludes std types. That implementation is also ready and in this PR.

So now it seems we just need to pick A or B.

Personally I think making std types work is what's ultimately better for *the users*. It's also needed for our own `cl::sycl::handler::fill` to work, since the "fill" function template uses the value type in its kernel name.

But either way, I wouldn't want this PR to keep going in circles, so... do we want A or B?

@rolandschulz @erichkeane @premanandrao @elizabethandrews

**Comment by erichkeane** [comment]
> 
> 
> Fixing the implementation so std names work seems a pointless exercise if we're going to block std names anyway with diagnostics. I'm trying to find the path that makes everyone happy. :-)
> 
> Several folks here expressed a strong desire to have std names work, and the good news is we _can_ make them work, and the implementation to do so -- templating KernelInfo on the string representation of the name -- is already in place; we just need to switch it on. No more dual modes.
> 
> Several folks here also expressed that the intention of the spec is that names and their template arguments be forward declarable, which excludes std types. That implementation is also ready and in this PR.
> 
> So now it seems we just need to pick A or B.
> 
> Personally I think making std types work is what's ultimately better for _the users_. It's also needed for our own `cl::sycl::handler::fill` to work, since the "fill" function template uses the value type in its kernel name.
> 
> But either way, I wouldn't want this PR to keep going in circles, so... do we want A or B?
> 
> @rolandschulz @erichkeane @premanandrao @elizabethandrews

Like I said, we cannot make the Unnamed Lambda extension default here.  As I've mentioned above, we want to disallow std types in NON unnamed lambda mode, and allow them in unnamed lambda mode (which is what @rolandschulz suggested a few weeks ago on 2271.  Basically, do your 'disallow std types' implementation, except don't disallow if unnamed lambdas is enabled.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2641]: Right, we shouldn't emit this diagnostic if UnnamedLambdaSupport

**Comment by jtmott-intel** [comment]
Roland also suggested "It seems like we could always do that: Template KernelInfo on the string representation of the name." :-) That's what I was following up on. But if that's a non-starter, then I can revert back.

The good news is I discovered that the entire "emit forward decls" call stack is already conditioned on `!UnnamedLambdaSupport` ([L#2766](https://github.com/intel/llvm/pull/1963/files#diff-c88a590aca6d5f8e92dd1f8b6ec05a49L2766)). I tested and verified that std names don't emit diagnostics with unnamed lambda option, so we should be good to go.

Now, though, it seems some tests are failing on Windows only. They report exit code 3221226505 and no stdout or stderr output. Googling the exit code suggests its an internal windows error. Those same tests passed earlier in this PR. :-/

**Comment by rolandschulz** [comment]
Lambda names are required in SYCL 1.2.1 but not in DPC++ or SYCL 2020 provisional. For the later the extension should be enabled by default. Also note that SYCL 1.2.1 doesn't require diagnostic for missing lambda names. Therefore in theory we could always enable the extension. But it would be good to have a warning for the users in SYCL 1.2.1 mode for them to know this isn't portable if they don't provide a lambda name. But such a diagnostic could be done independent of our implementation using the string representation (as we do for the extension) rather than the type (as we do without the extension).

**Comment by erichkeane** [comment]
> 
> 
> Lambda names are required in SYCL 1.2.1 but not in DPC++ or SYCL 2020 provisional. For the later the extension should be enabled by default. Also note that SYCL 1.2.1 doesn't require diagnostic for missing lambda names. Therefore in theory we could always enable the extension. But it would be good to have a warning for the users in SYCL 1.2.1 mode for them to know this isn't portable if they don't provide a lambda name. But such a diagnostic could be done independent of our implementation using the string representation (as we do for the extension) rather than the type (as we do without the extension).

I see, thanks for the update!  So we COULD still enable unnamed lambdas in SYCL 2020 mode.  

That said, we should still have the prohibition on std types in 1.2.1 mode.  I believe the enabling unnamed lambdas in 2020 mode is orthogonal to this patch, so I'd think we want to do that as a separate patch.

**Comment by jtmott-intel** [comment]
Hi, @bader . Any chance this PR is having a problem with the Windows buildbots? There's no significant code changes since the last "all checks passed". The specific tests that fail can be different on each run -- sometimes it's `reduction/reduction_transparent.cpp`, sometimes it's `host-interop-task/interop-task.cpp` -- but it's always Windows-only. And I haven't been able to reproduce the failures locally.

**Comment by romanovvlad** [comment]
> 
> 
> Hi, @bader . Any chance this PR is having a problem with the Windows buildbots? There's no significant code changes since the last "all checks passed". The specific tests that fail can be different on each run -- sometimes it's `reduction/reduction_transparent.cpp`, sometimes it's `host-interop-task/interop-task.cpp` -- but it's always Windows-only. And I haven't been able to reproduce the failures locally.

The test is flaky, we disabled it on windows. Could you please merge your branch with the latest sycl branch?

**Comment by jtmott-intel** [comment]
@romanovvlad After the latest merge from sycl, I'm getting a Windows-only error for `sycl/test/reduction/reduction_transparent.cpp` and `reduction_usm.cpp`. They're returning exit code 3221226505 and no output on stdout or stderr. I'm not familiar with that exit code but googling suggests it's a generic "Internal process error" coming from Windows. I can keep investigating but I wanted to check if this a buildbot issue?

**Comment by romanovvlad** [comment]
> 
> 
> @romanovvlad After the latest merge from sycl, I'm getting a Windows-only error for `sycl/test/reduction/reduction_transparent.cpp` and `reduction_usm.cpp`. They're returning exit code 3221226505 and no output on stdout or stderr. I'm not familiar with that exit code but googling suggests it's a generic "Internal process error" coming from Windows. I can keep investigating but I wanted to check if this a buildbot issue?

I have not seen this fails before. @vladimirlaz @tfzhu have you seen this issue in other PRs?
@jtmott-intel Are you sure that this issue is not introduced by your patch?

**Comment by jtmott-intel** [comment]
@romanovvlad This PR does touch the two failing tests, but I also got an "all checks passed" earlier, and since then I've only merged from sycl branch. I've looked over the changes to those files character by character, and everything still looks just as right as when all checks passed. None of the changes are OS-specific. And I've tried to reproduce the failure locally but can't reproduce.

**Comment by romanovvlad** [comment]
> 
> 
> @romanovvlad This PR does touch the two failing tests, but I also got an "all checks passed" earlier, and since then I've only merged from sycl branch. I've looked over the changes to those files character by character, and everything still looks just as right as when all checks passed. None of the changes are OS-specific. And I've tried to reproduce the failure locally but can't reproduce.

I've checked several other failing pre-commits on windows and I do not see these test failing.
Your patch may introduce sporadic failure.

**Comment by jtmott-intel** [comment]
All checks passed! Sorry, folks. Yes this latest problem was my fault. I made a mistake in a merge resolution and used the same kernel name twice. I didn't feel good about using a PR to troubleshoot the issue, but I couldn't reproduce it any other way, so I finally used a separate dummy PR to narrow down where the problem was coming from, and -- I'm happy to say it's fixed now. Final approval and merge may commence. I assume folks will prefer a squash merge for this one? :-)

@romanovvlad @elizabethandrews @premanandrao

**Review by premanandrao — APPROVED** [review]
@jtmott-intel, I don't see any major changes since I last reviewed this.   The set of changes look good to me.

@elizabethandrews, @Fznamznon, okay with you?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: I'm still a bit confused about this change. Can you give an example where `if (TD)` evaluates true and KernelNameIsMissing i.e. TD does not have a name? 

Also, this change is not technically required for 'std' namespace check right? Or am I missing something?

Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: I think this happens when it is invalid.
It is not needed strictly for the specific 'std' namespace check, but it occurs in those contexts too.

Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: Normally the `if (TD)` check would ensure that a kernel name type is a TagDecl -- that is, one of a struct/union/class/enum -- which would catch kernel name types such as `int`. I say "normally", though, because earlier in the call stack, we also ensure that a kernel name type is a CXXRecordDecl -- that is, one of a struct/union/class. Which means, sadly, that the `if (TD)` part may actually be dead code. **But**, it was dead code even before, and independent of, this PR. I had considered doing a clean-up of the existing code, but... I feel like this PR has already exceed its side-quest quota. :-)

**Comment by jtmott-intel** [comment]
I see @premanandrao 's approval, but github now appears to be blocking because it's waiting on an approval from someone in the [intel/llvm-reviewers-runtime team](https://github.com/orgs/intel/teams/llvm-reviewers-runtime/members). @romanovvlad @rbegam ?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: Sorry if I'm being obtuse here but it still isn't clear to me what this diagnostic is trying to catch. Can you give an example of the invalid code this diagnostic triggers for. Maybe then I'll understand why this diagnostic no longer needs to be guarded by UnnamedLambdaSupport.

**Review by rbegam — APPROVED** [review]
LGTM

File: clang/lib/Sema/SemaSYCL.cpp
Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: There's still an `UnnamedLambdaSupport` guard, just moved a few lines down, currently at L#1905. In the original pre-PR code, this condition -- `if (TD && !UnnamedLambdaSupport) {` -- had a \*potential\* problem because the "if" proceeds if `TD` is truthy (non-null), yet the "else" block still uses `TD` anyway.

I wrote "potential" problem because it all appears to be dead code anyway. Up the call stack, in `SYCLIntegrationHeader::emit` where we're writing out the source of the header, if there's no UnnamedLambdaSupport, then we don't call into `emitFwdDecl` at all. Which means, in that original pre-PR condition, `TD` would always be true, and `!UnnamedLambdaSupport` would also always be true, making the entire "else" block -- and the diagnositic within -- dead code.

@premanandrao Is there a specific case that you were fixing? Or was this general tidying?

Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: I don't believe it is dead-code; at least it wasn't when I refactored it.   I am pretty sure one of the cases in the test (as it was originally) would fail if this part were reverted.   So there is a simple way to check if this is still needed.   (i.e. Revert the refactoring and see if the unnamed-kernel.cpp test still passes.)

Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: Locally, I reverted the refactoring and still got passes for all the CodeGenSYCL and SemaSYCL tests. The refactoring commit was from back in April, so yes certainly lots of changes to the codebase since then. :-)

Is that the decision, then? Do we want to revert that? We'll have to wait a few hours for buildbot tests and then we'll all need to re-approve again.

Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: Fixing that code to not have useless dead code would be my preference.

Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: That would be my preference too **\*\*EXCEPT\*\*** this PR has already exceed its side-quest quota. :-) This PR doesn't introduce the dead code, which means we can merge *this* PR then open a separate one to clean up any dead code if we still want to do that. So the options I'm personally considering are: revert or don't revert the original refactoring from April.

Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: May be I am misunderstanding or not seeing the diffs correctly, are we talking about the:
```
const TagDecl *TD = isa<ClassTemplateDecl>(D)
                              ? cast<ClassTemplateDecl>(D)->getTemplatedDecl()
                              : dyn_cast<TagDecl>(D);
      if (!TD)
        break;
```
change?   If so, isn't that part of what is being checked in?

Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: Yes, but the original had dead code too. This change just moved it to a different line.

Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: I see.  In that case, I don't mind this being addressed in a separate PR.

Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp:2739]: OK then sounds like we're opting to *not* revert the April refactoring? And this PR would stay as is?

Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp]: Seems like folks still aren't comfortable with those changes, and since they come from an older time in this PR and don't seem to be needed anymore, I pushed a commit to revert that bit of refactoring. Buildbot is running. I'll ping folks when ready for re-review approvals.

Comment (jtmott-intel) [reply clang/lib/Sema/SemaSYCL.cpp]: The changes @elizabethandrews was concerned about are reverted. All checks passed. Need review approvals again. @premanandrao @rbegam

**Review by elizabethandrews — APPROVED** [review]
LGTM. Thanks! 

Can you file a GitHub issue to look into the dead code so we don't forget about it?

**Review by rbegam — APPROVED** [review]

**Comment by jtmott-intel** [comment]
The base branch restricts merging to authorized users -- and apparently I'm not an authorized user. :-) Maybe @premanandrao or @elizabethandrews can one of you click the merge button for me?

**Comment by elizabethandrews** [comment]
> The base branch restricts merging to authorized users -- and apparently I'm not an authorized user. :-) Maybe @premanandrao or @elizabethandrews can one of you click the merge button for me?

We're not authorized to either :) I see @bader already merged your commit but FYI, @bader and @pvchupin have merge rights.

**Pull request merged**
